### PR TITLE
Add ``tearDownForeignThreadGc`` function

### PR DESCRIPTION
**Pull request by vegansk**
The problem is that if we initialized GC using ``setupForeignThreadGc``, we get memory leaks on thread's exit. This PR adds function that should be used to avoid this.

### CODE REVIEW

**Comment by Araq** [comment]
Nice, with test case, I love it. :-)

**Comment by vegansk** [comment]
@Araq, this test shows that there is no public ``createThread`` definition in winlean, it's private and it's in ``threads`` module :-)

**Comment by vegansk** [comment]
There is one more upcoming change: we must ignore  ``setupForeignThreadGc`` and ``tearDownForeignThreadGc`` calls, if they are called inside the Nim's thread, main thread for example.

**Comment by Araq** [comment]
foreign_thr.nim is not run though, you need to patch testament/categories for this.

**Comment by vegansk** [comment]
@Araq, is it acceptable to check if GC is initialized with ``gch.stackBottom.isNil`` condition? If so, then PR is ready :-)

File: tests/gc/foreign_thr.nim [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    result = 0'i32
+
+  proc runInForeignThread(f: ThreadFunc) =
+    var dummyThreadId: DWORD
+    var h = createThread(nil, ThreadStackSize.int32, wrapper.WinThreadProc, cast[pointer](f), 0, dummyThreadId)
+    doAssert h != 0.Handle
+    doAssert waitForSingleObject(h, -1'i32) == 0.DWORD
+
+else:
+  {.fatal: "Unknown system".}
```
Comment (Araq) [inline tests/gc/foreign_thr.nim]: So on OSX this test will fail? Not good, just make it 'echo' things to make it green instead.

File: lib/system/gc_common.nim [hunk 2]
```diff
@@ -144,15 +144,35 @@ when allowForeignThreadGc:
     ## This function is available only when ``--threads:on`` and ``--tlsEmulation:off``
     ## switches are used
     if not localGcInitialized:
+      if not gch.stackBottom.isNil:
```
Comment (Araq) [inline lib/system/gc_common.nim]: This is too fragile. Why is it necessary? There is no guarantee that you can call setupForeignThreadGc in a Nim thread.

Comment (vegansk) [inline lib/system/gc_common.nim]: I have a dll written in Nim with some API. Each API function calls setupForeignThreadGc. But I have no guarantees that API will not be used in the main thread, that initialized via ``NimMain``, or in another Nim thread (via callback). We must allow to call ``setupForeignThreadGc`` and ``tearDownForeignThreadGc`` in the Nim thread.

File: tests/gc/foreign_thr.nim (hunk 1 above)
Comment (vegansk) [inline tests/gc/foreign_thr.nim]: Ok, then maybe it's better to use:
```nim
when defined(posix):
   ...
elif defined(windows):
   ...
else:
  {.fatal: "blah-blah".}

File: lib/system/gc_common.nim (hunk 2 above)
Comment (Araq) [inline lib/system/gc_common.nim]: well if you call it via callback, then this mechanism should wrap it in setup/tearDown.

Comment (vegansk) [inline lib/system/gc_common.nim]: remade this in another way

File: tests/gc/foreign_thr.nim (hunk 1 above)
Comment (Araq) [inline tests/gc/foreign_thr.nim]: sure

**Pull request merged**